fdbserver/BulkLoadUtil.actor.cpp

```diff
@@ -26,123 +26,92 @@
 #include "fdbserver/Knobs.h"
 #include "fdbserver/RocksDBCheckpointUtils.actor.h"
 #include "fdbserver/StorageMetrics.actor.h"
-#include "flow/Error.h"
-#include "flow/IAsyncFile.h"
-#include "flow/IRandom.h"
-#include "flow/Platform.h"
-#include "flow/Trace.h"
 #include "flow/genericactors.actor.h"
 #include "flow/actorcompiler.h" // has to be last include
 
-ACTOR Future<std::string> readBulkFileBytes(std::string path, int64_t maxLength) {
+ACTOR Future<Void> readBulkFileBytes(std::string path, int64_t maxLength, std::shared_ptr<std::string> output) {
 	try {
+		output->clear();
+		state int64_t chunkSize = SERVER_KNOBS->BULKLOAD_ASYNC_READ_WRITE_BLOCK_SIZE;
 		state Reference<IAsyncFile> file = wait(IAsyncFileSystem::filesystem()->open(
 		    abspath(path), IAsyncFile::OPEN_NO_AIO | IAsyncFile::OPEN_READONLY | IAsyncFile::OPEN_UNCACHED, 0644));
-
 		state int64_t fileSize = wait(file->size());
 		if (fileSize > maxLength) {
 			TraceEvent(SevError, "ReadBulkFileBytesTooLarge")
 			    .detail("FileSize", fileSize)
 			    .detail("MaxLength", maxLength);
 			throw file_too_large();
 		}
-
-		state std::shared_ptr<std::string> content = std::make_shared<std::string>();
-		content->reserve(fileSize); // Pre-allocate the full size
-
-		// For small files (< 1MB), do a single read
-		if (fileSize < 1024 * 1024) {
-			content->resize(fileSize);
-			int bytesRead = wait(uncancellable(holdWhile(content, file->read(content->data(), fileSize, 0))));
-			if (bytesRead != fileSize) {
+		output->reserve(fileSize); // Pre-allocate the full size
+
+		// Read in chunks to avoid memory pressure
+		state int64_t offset = 0;
+		state int64_t remaining = fileSize;
+		while (remaining > 0) {
+			state int64_t bytesToRead = std::min(chunkSize, remaining);
+			state std::shared_ptr<std::string> chunk = std::make_shared<std::string>();
+			chunk->resize(bytesToRead);
```

Comment:
Why allocate each time through the loop?

Reply:
Will fix this in the next PR.